// @wordpress/block-editor/build-module/components/link-control/search-input.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { useState } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { Button } from '@wordpress/components';
import { LEFT, RIGHT, UP, DOWN, BACKSPACE, ENTER } from '@wordpress/keycodes';
/**
 * Internal dependencies
 */

import { URLInput } from '../';

var handleLinkControlOnKeyDown = function handleLinkControlOnKeyDown(event) {
  var keyCode = event.keyCode;

  if ([LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER].indexOf(keyCode) > -1) {
    // Stop the key event from propagating up to ObserveTyping.startTypingInTextField.
    event.stopPropagation();
  }
};

var handleLinkControlOnKeyPress = function handleLinkControlOnKeyPress(event) {
  var keyCode = event.keyCode;
  event.stopPropagation();

  if (keyCode === ENTER) {}
};

var LinkControlSearchInput = function LinkControlSearchInput(_ref) {
  var value = _ref.value,
      onChange = _ref.onChange,
      onSelect = _ref.onSelect,
      renderSuggestions = _ref.renderSuggestions,
      fetchSuggestions = _ref.fetchSuggestions,
      showInitialSuggestions = _ref.showInitialSuggestions;

  var _useState = useState(),
      _useState2 = _slicedToArray(_useState, 2),
      selectedSuggestion = _useState2[0],
      setSelectedSuggestion = _useState2[1];

  var selectItemHandler = function selectItemHandler(selection, suggestion) {
    onChange(selection);
    setSelectedSuggestion(suggestion);
  };

  function selectSuggestionOrCurrentInputValue(event) {
    // Avoid default forms behavior, since it's being handled custom here.
    event.preventDefault(); // Interpret the selected value as either the selected suggestion, if
    // exists, or otherwise the current input value as entered.

    onSelect(selectedSuggestion || {
      url: value
    });
  }

  return createElement("form", {
    onSubmit: selectSuggestionOrCurrentInputValue
  }, createElement(URLInput, {
    className: "block-editor-link-control__search-input",
    value: value,
    onChange: selectItemHandler,
    onKeyDown: function onKeyDown(event) {
      if (event.keyCode === ENTER) {
        return;
      }

      handleLinkControlOnKeyDown(event);
    },
    onKeyPress: handleLinkControlOnKeyPress,
    placeholder: __('Search or type url'),
    __experimentalRenderSuggestions: renderSuggestions,
    __experimentalFetchLinkSuggestions: fetchSuggestions,
    __experimentalHandleURLSuggestions: true,
    __experimentalShowInitialSuggestions: showInitialSuggestions
  }), createElement("div", {
    className: "block-editor-link-control__search-actions"
  }, createElement(Button, {
    type: "submit",
    label: __('Submit'),
    icon: "editor-break",
    className: "block-editor-link-control__search-submit"
  })));
};

export default LinkControlSearchInput;
