// @wordpress/block-editor/build/components/page-template-picker/picker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _blocks = require("@wordpress/blocks");

var _data = require("@wordpress/data");

var _button = _interopRequireDefault(require("./button"));

var _container = _interopRequireDefault(require("./container"));

var _defaultTemplates = _interopRequireDefault(require("./default-templates"));

var _preview = _interopRequireDefault(require("./preview"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var __experimentalPageTemplatePicker = function __experimentalPageTemplatePicker(_ref) {
  var _ref$templates = _ref.templates,
      templates = _ref$templates === void 0 ? (0, _defaultTemplates.default)() : _ref$templates;

  var _useDispatch = (0, _data.useDispatch)('core/editor'),
      editPost = _useDispatch.editPost;

  var _useState = (0, _element.useState)(),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      templatePreview = _useState2[0],
      setTemplatePreview = _useState2[1];

  var onApply = function onApply() {
    editPost({
      title: templatePreview.name,
      blocks: (0, _blocks.parse)(templatePreview.content)
    });
    setTemplatePreview(undefined);
  };

  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_container.default, null, templates.map(function (template) {
    return (0, _element.createElement)(_button.default, {
      key: template.name,
      icon: template.icon,
      label: template.name,
      onPress: function onPress() {
        return setTemplatePreview(template);
      }
    });
  })), (0, _element.createElement)(_preview.default, {
    template: templatePreview,
    onDismiss: function onDismiss() {
      return setTemplatePreview(undefined);
    },
    onApply: onApply
  }));
};

var _default = __experimentalPageTemplatePicker;
exports.default = _default;
