// @wordpress/block-editor/build/components/navigable-toolbar/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _components = require("@wordpress/components");

var _dom = require("@wordpress/dom");

var _keyboardShortcuts = require("@wordpress/keyboard-shortcuts");

/**
 * WordPress dependencies
 */
function NavigableToolbar(_ref) {
  var children = _ref.children,
      focusOnMount = _ref.focusOnMount,
      props = (0, _objectWithoutProperties2.default)(_ref, ["children", "focusOnMount"]);
  var wrapper = (0, _element.useRef)();
  var focusToolbar = (0, _element.useCallback)(function () {
    var tabbables = _dom.focus.tabbable.find(wrapper.current);

    if (tabbables.length) {
      tabbables[0].focus();
    }
  }, []);
  (0, _keyboardShortcuts.useShortcut)('core/block-editor/focus-toolbar', focusToolbar, {
    bindGlobal: true,
    eventName: 'keydown'
  });
  (0, _element.useEffect)(function () {
    if (focusOnMount) {
      focusToolbar();
    }
  }, []);
  return (0, _element.createElement)(_components.NavigableMenu, (0, _extends2.default)({
    orientation: "horizontal",
    role: "toolbar",
    ref: wrapper
  }, props), children);
}

var _default = NavigableToolbar;
exports.default = _default;
