// @wordpress/block-editor/build-module/components/block-mover/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { first, last, partial, castArray } from 'lodash';
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
import { Button, ToolbarGroup } from '@wordpress/components';
import { getBlockType } from '@wordpress/blocks';
import { Component } from '@wordpress/element';
import { withSelect, withDispatch } from '@wordpress/data';
import { withInstanceId, compose } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import { getBlockMoverDescription } from './mover-description';
import { leftArrow, rightArrow, dragHandle } from './icons';
import { chevronUp, chevronDown } from '@wordpress/icons';
import BlockDraggable from '../block-draggable';
export var BlockMover =
/*#__PURE__*/
function (_Component) {
  _inherits(BlockMover, _Component);

  function BlockMover() {
    var _this;

    _classCallCheck(this, BlockMover);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(BlockMover).apply(this, arguments));
    _this.state = {
      isFocused: false
    };
    _this.onFocus = _this.onFocus.bind(_assertThisInitialized(_this));
    _this.onBlur = _this.onBlur.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(BlockMover, [{
    key: "onFocus",
    value: function onFocus() {
      this.setState({
        isFocused: true
      });
    }
  }, {
    key: "onBlur",
    value: function onBlur() {
      this.setState({
        isFocused: false
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          onMoveUp = _this$props.onMoveUp,
          onMoveDown = _this$props.onMoveDown,
          orientation = _this$props.__experimentalOrientation,
          isRTL = _this$props.isRTL,
          isFirst = _this$props.isFirst,
          isLast = _this$props.isLast,
          clientIds = _this$props.clientIds,
          blockType = _this$props.blockType,
          firstIndex = _this$props.firstIndex,
          isLocked = _this$props.isLocked,
          instanceId = _this$props.instanceId,
          isHidden = _this$props.isHidden,
          rootClientId = _this$props.rootClientId,
          hideDragHandle = _this$props.hideDragHandle;
      var isFocused = this.state.isFocused;
      var blocksCount = castArray(clientIds).length;

      if (isLocked || isFirst && isLast && !rootClientId) {
        return null;
      }

      var getArrowIcon = function getArrowIcon(moveDirection) {
        if (moveDirection === 'up') {
          if (orientation === 'horizontal') {
            return isRTL ? rightArrow : leftArrow;
          }

          return chevronUp;
        } else if (moveDirection === 'down') {
          if (orientation === 'horizontal') {
            return isRTL ? leftArrow : rightArrow;
          }

          return chevronDown;
        }

        return null;
      };

      var getMovementDirectionLabel = function getMovementDirectionLabel(moveDirection) {
        if (moveDirection === 'up') {
          if (orientation === 'horizontal') {
            return isRTL ? __('Move right') : __('Move left');
          }

          return __('Move up');
        } else if (moveDirection === 'down') {
          if (orientation === 'horizontal') {
            return isRTL ? __('Move left') : __('Move right');
          }

          return __('Move down');
        }

        return null;
      }; // We emulate a disabled state because forcefully applying the `disabled`
      // attribute on the button while it has focus causes the screen to change
      // to an unfocused state (body as active element) without firing blur on,
      // the rendering parent, leaving it unable to react to focus out.


      return createElement(ToolbarGroup, {
        className: classnames('block-editor-block-mover', {
          'is-visible': isFocused || !isHidden,
          'is-horizontal': orientation === 'horizontal'
        })
      }, createElement(Button, {
        className: "block-editor-block-mover__control",
        onClick: isFirst ? null : onMoveUp,
        icon: getArrowIcon('up'),
        label: getMovementDirectionLabel('up'),
        "aria-describedby": "block-editor-block-mover__up-description-".concat(instanceId),
        "aria-disabled": isFirst,
        onFocus: this.onFocus,
        onBlur: this.onBlur
      }), !hideDragHandle && createElement(BlockDraggable, {
        clientIds: clientIds
      }, function (_ref) {
        var onDraggableStart = _ref.onDraggableStart,
            onDraggableEnd = _ref.onDraggableEnd;
        return createElement(Button, {
          icon: dragHandle,
          className: "block-editor-block-mover__control-drag-handle block-editor-block-mover__control",
          "aria-hidden": "true" // Should not be able to tab to drag handle as this
          // button can only be used with a pointer device.
          ,
          tabIndex: "-1",
          onDragStart: onDraggableStart,
          onDragEnd: onDraggableEnd,
          draggable: true
        });
      }), createElement(Button, {
        className: "block-editor-block-mover__control",
        onClick: isLast ? null : onMoveDown,
        icon: getArrowIcon('down'),
        label: getMovementDirectionLabel('down'),
        "aria-describedby": "block-editor-block-mover__down-description-".concat(instanceId),
        "aria-disabled": isLast,
        onFocus: this.onFocus,
        onBlur: this.onBlur
      }), createElement("span", {
        id: "block-editor-block-mover__up-description-".concat(instanceId),
        className: "block-editor-block-mover__description"
      }, getBlockMoverDescription(blocksCount, blockType && blockType.title, firstIndex, isFirst, isLast, -1, orientation, isRTL)), createElement("span", {
        id: "block-editor-block-mover__down-description-".concat(instanceId),
        className: "block-editor-block-mover__description"
      }, getBlockMoverDescription(blocksCount, blockType && blockType.title, firstIndex, isFirst, isLast, 1, orientation, isRTL)));
    }
  }]);

  return BlockMover;
}(Component);
export default compose(withSelect(function (select, _ref2) {
  var clientIds = _ref2.clientIds;

  var _select = select('core/block-editor'),
      getBlock = _select.getBlock,
      getBlockIndex = _select.getBlockIndex,
      getTemplateLock = _select.getTemplateLock,
      getBlockRootClientId = _select.getBlockRootClientId,
      getBlockOrder = _select.getBlockOrder;

  var normalizedClientIds = castArray(clientIds);
  var firstClientId = first(normalizedClientIds);
  var block = getBlock(firstClientId);
  var rootClientId = getBlockRootClientId(first(normalizedClientIds));
  var blockOrder = getBlockOrder(rootClientId);
  var firstIndex = getBlockIndex(firstClientId, rootClientId);
  var lastIndex = getBlockIndex(last(normalizedClientIds), rootClientId);

  var _select2 = select('core/block-editor'),
      getSettings = _select2.getSettings;

  var _getSettings = getSettings(),
      isRTL = _getSettings.isRTL;

  return {
    blockType: block ? getBlockType(block.name) : null,
    isLocked: getTemplateLock(rootClientId) === 'all',
    rootClientId: rootClientId,
    firstIndex: firstIndex,
    isRTL: isRTL,
    isFirst: firstIndex === 0,
    isLast: lastIndex === blockOrder.length - 1
  };
}), withDispatch(function (dispatch, _ref3) {
  var clientIds = _ref3.clientIds,
      rootClientId = _ref3.rootClientId;

  var _dispatch = dispatch('core/block-editor'),
      moveBlocksDown = _dispatch.moveBlocksDown,
      moveBlocksUp = _dispatch.moveBlocksUp;

  return {
    onMoveDown: partial(moveBlocksDown, clientIds, rootClientId),
    onMoveUp: partial(moveBlocksUp, clientIds, rootClientId)
  };
}), withInstanceId)(BlockMover);
