// @wordpress/block-editor/build-module/components/page-template-picker/button.native.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { TouchableOpacity, Text } from 'react-native';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
import { withPreferredColorScheme } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import styles from './styles.scss';

var PickerButton = function PickerButton(_ref) {
  var icon = _ref.icon,
      label = _ref.label,
      onPress = _ref.onPress,
      getStylesFromColorScheme = _ref.getStylesFromColorScheme;
  var butonStyles = getStylesFromColorScheme(styles.button, styles.buttonDark);
  var butonTextStyles = getStylesFromColorScheme(styles.buttonText, styles.buttonTextDark);
  return createElement(TouchableOpacity, {
    accessibilityLabel: label,
    accessibilityHint: __('Double tap to select layout'),
    activeOpacity: 0.7,
    onPress: onPress,
    style: butonStyles
  }, createElement(Text, {
    style: styles.buttonIcon
  }, icon), createElement(Text, {
    style: butonTextStyles
  }, label));
};

export default withPreferredColorScheme(PickerButton);
