// @wordpress/block-editor/build/components/tool-selector/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _compose = require("@wordpress/compose");

/**
 * WordPress dependencies
 */
var editIcon = (0, _element.createElement)(_components.SVG, {
  xmlns: "http://www.w3.org/2000/svg",
  width: "20",
  height: "20",
  viewBox: "0 0 24 24"
}, (0, _element.createElement)(_components.Path, {
  fill: "none",
  d: "M0 0h24v24H0V0z"
}), (0, _element.createElement)(_components.Path, {
  d: "M14.06 9.02l.92.92L5.92 19H5v-.92l9.06-9.06M17.66 3c-.25 0-.51.1-.7.29l-1.83 1.83 3.75 3.75 1.83-1.83c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.2-.2-.45-.29-.71-.29zm-3.6 3.19L3 17.25V21h3.75L17.81 9.94l-3.75-3.75z"
}));
var selectIcon = (0, _element.createElement)(_components.SVG, {
  xmlns: "http://www.w3.org/2000/svg",
  width: "20",
  height: "20",
  viewBox: "0 0 24 24"
}, (0, _element.createElement)(_components.Path, {
  d: "M6.5 1v21.5l6-6.5H21L6.5 1zm5.1 13l-3.1 3.4V5.9l7.8 8.1h-4.7z"
}));

function ToolSelector() {
  var isNavigationTool = (0, _data.useSelect)(function (select) {
    return select('core/block-editor').isNavigationMode();
  }, []);

  var _useDispatch = (0, _data.useDispatch)('core/block-editor'),
      setNavigationMode = _useDispatch.setNavigationMode;

  var isMediumViewport = (0, _compose.useViewportMatch)('medium');

  if (!isMediumViewport) {
    return null;
  }

  var onSwitchMode = function onSwitchMode(mode) {
    setNavigationMode(mode === 'edit' ? false : true);
  };

  return (0, _element.createElement)(_components.Dropdown, {
    renderToggle: function renderToggle(_ref) {
      var isOpen = _ref.isOpen,
          onToggle = _ref.onToggle;
      return (0, _element.createElement)(_components.Button, {
        icon: isNavigationTool ? selectIcon : editIcon,
        "aria-expanded": isOpen,
        onClick: onToggle,
        label: (0, _i18n.__)('Tools')
      });
    },
    renderContent: function renderContent() {
      return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_components.NavigableMenu, {
        role: "menu",
        "aria-label": (0, _i18n.__)('Tools')
      }, (0, _element.createElement)(_components.MenuItemsChoice, {
        value: isNavigationTool ? 'select' : 'edit',
        onSelect: onSwitchMode,
        choices: [{
          value: 'edit',
          label: (0, _element.createElement)(_element.Fragment, null, editIcon, (0, _i18n.__)('Edit'))
        }, {
          value: 'select',
          label: (0, _element.createElement)(_element.Fragment, null, selectIcon, (0, _i18n.__)('Select'))
        }]
      })), (0, _element.createElement)("div", {
        className: "block-editor-tool-selector__help"
      }, (0, _i18n.__)('Tools offer different interactions for block selection & editing. To select, press Escape, to go back to editing, press Enter.')));
    }
  });
}

var _default = ToolSelector;
exports.default = _default;
