// @wordpress/block-editor/build/components/colors/use-colors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = __experimentalUseColors;

var _element = require("@wordpress/element");

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _memize = _interopRequireDefault(require("memize"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _lodash = require("lodash");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _panelColorSettings = _interopRequireDefault(require("../panel-color-settings"));

var _contrastChecker = _interopRequireDefault(require("../contrast-checker"));

var _inspectorControls = _interopRequireDefault(require("../inspector-controls"));

var _blockEdit = require("../block-edit");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * Browser dependencies
 */
var _window = window,
    getComputedStyle = _window.getComputedStyle,
    Node = _window.Node;
var DEFAULT_COLORS = [];
var COMMON_COLOR_LABELS = {
  textColor: (0, _i18n.__)('Text Color'),
  backgroundColor: (0, _i18n.__)('Background Color')
};

var resolveContrastCheckerColor = function resolveContrastCheckerColor(color, colorSettings, detectedColor) {
  if (typeof color === 'function') {
    return color(colorSettings);
  } else if (color === true) {
    return detectedColor;
  }

  return color;
};

var ColorPanel = function ColorPanel(_ref) {
  var title = _ref.title,
      colorSettings = _ref.colorSettings,
      colorPanelProps = _ref.colorPanelProps,
      contrastCheckers = _ref.contrastCheckers,
      detectedBackgroundColor = _ref.detectedBackgroundColor,
      detectedColor = _ref.detectedColor,
      panelChildren = _ref.panelChildren,
      initialOpen = _ref.initialOpen;
  return (0, _element.createElement)(_panelColorSettings.default, (0, _extends2.default)({
    title: title,
    initialOpen: initialOpen,
    colorSettings: Object.values(colorSettings)
  }, colorPanelProps), contrastCheckers && (Array.isArray(contrastCheckers) ? contrastCheckers.map(function (_ref2) {
    var backgroundColor = _ref2.backgroundColor,
        textColor = _ref2.textColor,
        rest = (0, _objectWithoutProperties2.default)(_ref2, ["backgroundColor", "textColor"]);
    backgroundColor = resolveContrastCheckerColor(backgroundColor, colorSettings, detectedBackgroundColor);
    textColor = resolveContrastCheckerColor(textColor, colorSettings, detectedColor);
    return (0, _element.createElement)(_contrastChecker.default, (0, _extends2.default)({
      key: "".concat(backgroundColor, "-").concat(textColor),
      backgroundColor: backgroundColor,
      textColor: textColor
    }, rest));
  }) : (0, _lodash.map)(colorSettings, function (_ref3) {
    var value = _ref3.value;
    var backgroundColor = contrastCheckers.backgroundColor,
        textColor = contrastCheckers.textColor;
    backgroundColor = resolveContrastCheckerColor(backgroundColor || value, colorSettings, detectedBackgroundColor);
    textColor = resolveContrastCheckerColor(textColor || value, colorSettings, detectedColor);
    return (0, _element.createElement)(_contrastChecker.default, (0, _extends2.default)({}, contrastCheckers, {
      key: "".concat(backgroundColor, "-").concat(textColor),
      backgroundColor: backgroundColor,
      textColor: textColor
    }));
  })), typeof panelChildren === 'function' ? panelChildren(colorSettings) : panelChildren);
};

var InspectorControlsColorPanel = function InspectorControlsColorPanel(props) {
  return (0, _element.createElement)(_inspectorControls.default, null, (0, _element.createElement)(ColorPanel, props));
};

function __experimentalUseColors(colorConfigs) {
  var _ref4 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
    panelTitle: (0, _i18n.__)('Color settings')
  },
      _ref4$panelTitle = _ref4.panelTitle,
      panelTitle = _ref4$panelTitle === void 0 ? (0, _i18n.__)('Color settings') : _ref4$panelTitle,
      colorPanelProps = _ref4.colorPanelProps,
      contrastCheckers = _ref4.contrastCheckers,
      panelChildren = _ref4.panelChildren,
      _ref4$colorDetector = _ref4.colorDetector;

  _ref4$colorDetector = _ref4$colorDetector === void 0 ? {} : _ref4$colorDetector;
  var targetRef = _ref4$colorDetector.targetRef,
      _ref4$colorDetector$b = _ref4$colorDetector.backgroundColorTargetRef,
      backgroundColorTargetRef = _ref4$colorDetector$b === void 0 ? targetRef : _ref4$colorDetector$b,
      _ref4$colorDetector$t = _ref4$colorDetector.textColorTargetRef,
      textColorTargetRef = _ref4$colorDetector$t === void 0 ? targetRef : _ref4$colorDetector$t;
  var deps = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];

  var _useBlockEditContext = (0, _blockEdit.useBlockEditContext)(),
      clientId = _useBlockEditContext.clientId;

  var _useSelect = (0, _data.useSelect)(function (select) {
    var _select = select('core/block-editor'),
        getBlockAttributes = _select.getBlockAttributes,
        getSettings = _select.getSettings;

    var colors = getSettings().colors;
    return {
      attributes: getBlockAttributes(clientId),
      settingsColors: !colors || colors === true ? DEFAULT_COLORS : colors
    };
  }, [clientId]),
      attributes = _useSelect.attributes,
      settingsColors = _useSelect.settingsColors;

  var _useDispatch = (0, _data.useDispatch)('core/block-editor'),
      updateBlockAttributes = _useDispatch.updateBlockAttributes;

  var setAttributes = (0, _element.useCallback)(function (newAttributes) {
    return updateBlockAttributes(clientId, newAttributes);
  }, [updateBlockAttributes, clientId]);
  var createComponent = (0, _element.useMemo)(function () {
    return (0, _memize.default)(function (name, property, className, color, colorValue, customColor) {
      return function (_ref5) {
        var children = _ref5.children,
            _ref5$className = _ref5.className,
            componentClassName = _ref5$className === void 0 ? '' : _ref5$className,
            _ref5$style = _ref5.style,
            componentStyle = _ref5$style === void 0 ? {} : _ref5$style;
        return (// Clone children, setting the style property from the color configuration,
          // if not already set explicitly through props.
          _element.Children.map(children, function (child) {
            var _classnames;

            var colorStyle = {};

            if (color) {
              colorStyle = (0, _defineProperty2.default)({}, property, colorValue);
            } else if (customColor) {
              colorStyle = (0, _defineProperty2.default)({}, property, customColor);
            }

            return (0, _element.cloneElement)(child, {
              className: (0, _classnames2.default)(componentClassName, child.props.className, (_classnames = {}, (0, _defineProperty2.default)(_classnames, "has-".concat((0, _lodash.kebabCase)(color), "-").concat((0, _lodash.kebabCase)(property)), color), (0, _defineProperty2.default)(_classnames, className || "has-".concat((0, _lodash.kebabCase)(name)), color || customColor), _classnames)),
              style: _objectSpread({}, colorStyle, {}, componentStyle, {}, child.props.style || {})
            });
          })
        );
      };
    }, {
      maxSize: colorConfigs.length
    });
  }, [colorConfigs.length]);
  var createSetColor = (0, _element.useMemo)(function () {
    return (0, _memize.default)(function (name, colors) {
      return function (newColor) {
        var color = colors.find(function (_color) {
          return _color.color === newColor;
        });
        setAttributes((0, _defineProperty2.default)({}, color ? (0, _lodash.camelCase)("custom ".concat(name)) : name, undefined));
        setAttributes((0, _defineProperty2.default)({}, color ? name : (0, _lodash.camelCase)("custom ".concat(name)), color ? color.slug : newColor));
      };
    }, {
      maxSize: colorConfigs.length
    });
  }, [setAttributes, colorConfigs.length]);

  var _useState = (0, _element.useState)(),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      detectedBackgroundColor = _useState2[0],
      setDetectedBackgroundColor = _useState2[1];

  var _useState3 = (0, _element.useState)(),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      detectedColor = _useState4[0],
      setDetectedColor = _useState4[1];

  (0, _element.useEffect)(function () {
    if (!contrastCheckers) {
      return undefined;
    }

    var needsBackgroundColor = false;
    var needsColor = false;
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = (0, _lodash.castArray)(contrastCheckers)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var _step$value = _step.value,
            _backgroundColor = _step$value.backgroundColor,
            textColor = _step$value.textColor;

        if (!needsBackgroundColor) {
          needsBackgroundColor = _backgroundColor === true;
        }

        if (!needsColor) {
          needsColor = textColor === true;
        }

        if (needsBackgroundColor && needsColor) {
          break;
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    if (needsColor) {
      setDetectedColor(getComputedStyle(textColorTargetRef.current).color);
    }

    if (needsBackgroundColor) {
      var backgroundColorNode = backgroundColorTargetRef.current;
      var backgroundColor = getComputedStyle(backgroundColorNode).backgroundColor;

      while (backgroundColor === 'rgba(0, 0, 0, 0)' && backgroundColorNode.parentNode && backgroundColorNode.parentNode.nodeType === Node.ELEMENT_NODE) {
        backgroundColorNode = backgroundColorNode.parentNode;
        backgroundColor = getComputedStyle(backgroundColorNode).backgroundColor;
      }

      setDetectedBackgroundColor(backgroundColor);
    }
  }, [colorConfigs.reduce(function (acc, colorConfig) {
    return "".concat(acc, " | ").concat(attributes[colorConfig.name], " | ").concat(attributes[(0, _lodash.camelCase)("custom ".concat(colorConfig.name))]);
  }, '')].concat((0, _toConsumableArray2.default)(deps)));
  return (0, _element.useMemo)(function () {
    var colorSettings = {};
    var components = colorConfigs.reduce(function (acc, colorConfig) {
      if (typeof colorConfig === 'string') {
        colorConfig = {
          name: colorConfig
        };
      }

      var _colorConfig$color = _objectSpread({}, colorConfig, {
        color: attributes[colorConfig.name]
      }),
          name = _colorConfig$color.name,
          _colorConfig$color$pr = _colorConfig$color.property,
          property = _colorConfig$color$pr === void 0 ? name : _colorConfig$color$pr,
          className = _colorConfig$color.className,
          _colorConfig$color$pa = _colorConfig$color.panelLabel,
          panelLabel = _colorConfig$color$pa === void 0 ? colorConfig.label || COMMON_COLOR_LABELS[name] || (0, _lodash.startCase)(name) : _colorConfig$color$pa,
          _colorConfig$color$co = _colorConfig$color.componentName,
          componentName = _colorConfig$color$co === void 0 ? (0, _lodash.startCase)(name).replace(/\s/g, '') : _colorConfig$color$co,
          _colorConfig$color$co2 = _colorConfig$color.color,
          color = _colorConfig$color$co2 === void 0 ? colorConfig.color : _colorConfig$color$co2,
          _colorConfig$color$co3 = _colorConfig$color.colors,
          colors = _colorConfig$color$co3 === void 0 ? settingsColors : _colorConfig$color$co3;

      var customColor = attributes[(0, _lodash.camelCase)("custom ".concat(name))]; // We memoize the non-primitives to avoid unnecessary updates
      // when they are used as props for other components.

      var _color = customColor ? undefined : colors.find(function (__color) {
        return __color.slug === color;
      });

      acc[componentName] = createComponent(name, property, className, color, _color && _color.color, customColor);
      acc[componentName].displayName = componentName;
      acc[componentName].color = customColor ? customColor : _color && _color.color;
      acc[componentName].slug = color;
      acc[componentName].setColor = createSetColor(name, colors);
      colorSettings[componentName] = {
        value: _color ? _color.color : attributes[(0, _lodash.camelCase)("custom ".concat(name))],
        onChange: acc[componentName].setColor,
        label: panelLabel,
        colors: colors
      }; // These settings will be spread over the `colors` in
      // `colorPanelProps`, so we need to unset the key here,
      // if not set to an actual value, to avoid overwriting
      // an actual value in `colorPanelProps`.

      if (!colors) {
        delete colorSettings[componentName].colors;
      }

      return acc;
    }, {});
    var wrappedColorPanelProps = {
      title: panelTitle,
      initialOpen: false,
      colorSettings: colorSettings,
      colorPanelProps: colorPanelProps,
      contrastCheckers: contrastCheckers,
      detectedBackgroundColor: detectedBackgroundColor,
      detectedColor: detectedColor,
      panelChildren: panelChildren
    };
    return _objectSpread({}, components, {
      ColorPanel: (0, _element.createElement)(ColorPanel, wrappedColorPanelProps),
      InspectorControlsColorPanel: (0, _element.createElement)(InspectorControlsColorPanel, wrappedColorPanelProps)
    });
  }, [attributes, setAttributes, detectedColor, detectedBackgroundColor].concat((0, _toConsumableArray2.default)(deps)));
}
