// @wordpress/block-editor/build-module/components/button-block-appender/index.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { Button, Tooltip } from '@wordpress/components';
import { _x, sprintf } from '@wordpress/i18n';
import { Icon, plusCircle } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import Inserter from '../inserter';

function ButtonBlockAppender(_ref) {
  var rootClientId = _ref.rootClientId,
      className = _ref.className,
      selectBlockOnInsert = _ref.__experimentalSelectBlockOnInsert;
  return createElement(Inserter, {
    rootClientId: rootClientId,
    __experimentalSelectBlockOnInsert: selectBlockOnInsert,
    renderToggle: function renderToggle(_ref2) {
      var onToggle = _ref2.onToggle,
          disabled = _ref2.disabled,
          isOpen = _ref2.isOpen,
          blockTitle = _ref2.blockTitle,
          hasSingleBlockType = _ref2.hasSingleBlockType;
      var label;

      if (hasSingleBlockType) {
        // translators: %s: the name of the block when there is only one
        label = sprintf(_x('Add %s', 'directly add the only allowed block'), blockTitle);
      } else {
        label = _x('Add block', 'Generic label for block inserter button');
      }

      var isToggleButton = !hasSingleBlockType;
      return createElement(Tooltip, {
        text: label
      }, createElement(Button, {
        className: classnames(className, 'block-editor-button-block-appender'),
        onClick: onToggle,
        "aria-haspopup": isToggleButton ? 'true' : undefined,
        "aria-expanded": isToggleButton ? isOpen : undefined,
        disabled: disabled,
        label: label
      }, createElement("span", {
        className: "screen-reader-text"
      }, label), createElement(Icon, {
        icon: plusCircle
      })));
    },
    isAppender: true
  });
}
/**
 * @see https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/button-block-appender/README.md
 */


export default ButtonBlockAppender;
