// @wordpress/block-editor/build/components/block-navigation/list.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BlockNavigationList;

var _element = require("@wordpress/element");

var _lodash = require("lodash");

var _classnames = _interopRequireDefault(require("classnames"));

var _components = require("@wordpress/components");

var _blocks = require("@wordpress/blocks");

var _i18n = require("@wordpress/i18n");

var _blockIcon = _interopRequireDefault(require("../block-icon"));

var _buttonBlockAppender = _interopRequireDefault(require("../button-block-appender"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function BlockNavigationList(_ref) {
  var blocks = _ref.blocks,
      selectedBlockClientId = _ref.selectedBlockClientId,
      selectBlock = _ref.selectBlock,
      showAppender = _ref.showAppender,
      showNestedBlocks = _ref.showNestedBlocks,
      parentBlockClientId = _ref.parentBlockClientId;
  var shouldShowAppender = showAppender && !!parentBlockClientId;
  return (
    /*
     * Disable reason: The `list` ARIA role is redundant but
     * Safari+VoiceOver won't announce the list otherwise.
     */

    /* eslint-disable jsx-a11y/no-redundant-roles */
    (0, _element.createElement)("ul", {
      className: "block-editor-block-navigation__list",
      role: "list"
    }, (0, _lodash.map)((0, _lodash.omitBy)(blocks, _lodash.isNil), function (block) {
      var blockType = (0, _blocks.getBlockType)(block.name);
      var isSelected = block.clientId === selectedBlockClientId;
      return (0, _element.createElement)("li", {
        key: block.clientId
      }, (0, _element.createElement)("div", {
        className: "block-editor-block-navigation__item"
      }, (0, _element.createElement)(_components.Button, {
        className: (0, _classnames.default)('block-editor-block-navigation__item-button', {
          'is-selected': isSelected
        }),
        onClick: function onClick() {
          return selectBlock(block.clientId);
        }
      }, (0, _element.createElement)(_blockIcon.default, {
        icon: blockType.icon,
        showColors: true
      }), (0, _blocks.__experimentalGetBlockLabel)(blockType, block.attributes), isSelected && (0, _element.createElement)("span", {
        className: "screen-reader-text"
      }, (0, _i18n.__)('(selected block)')))), showNestedBlocks && !!block.innerBlocks && !!block.innerBlocks.length && (0, _element.createElement)(BlockNavigationList, {
        blocks: block.innerBlocks,
        selectedBlockClientId: selectedBlockClientId,
        selectBlock: selectBlock,
        parentBlockClientId: block.clientId,
        showAppender: showAppender,
        showNestedBlocks: true
      }));
    }), shouldShowAppender && (0, _element.createElement)("li", null, (0, _element.createElement)("div", {
      className: "block-editor-block-navigation__item"
    }, (0, _element.createElement)(_buttonBlockAppender.default, {
      rootClientId: parentBlockClientId,
      __experimentalSelectBlockOnInsert: false
    }))))
    /* eslint-enable jsx-a11y/no-redundant-roles */

  );
}
