// @wordpress/block-editor/build-module/components/editor-skeleton/index.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { useEffect } from '@wordpress/element';
import { navigateRegions } from '@wordpress/components';
import { __ } from '@wordpress/i18n';

function useHTMLClass(className) {
  useEffect(function () {
    var element = document && document.querySelector("html:not(.".concat(className, ")"));

    if (!element) {
      return;
    }

    element.classList.toggle(className);
    return function () {
      element.classList.toggle(className);
    };
  }, [className]);
}

function EditorSkeleton(_ref) {
  var footer = _ref.footer,
      header = _ref.header,
      sidebar = _ref.sidebar,
      content = _ref.content,
      publish = _ref.publish,
      className = _ref.className;
  useHTMLClass('block-editor-editor-skeleton__html-container');
  return createElement("div", {
    className: classnames(className, 'block-editor-editor-skeleton')
  }, !!header && createElement("div", {
    className: "block-editor-editor-skeleton__header",
    role: "region"
    /* translators: accessibility text for the top bar landmark region. */
    ,
    "aria-label": __('Editor top bar'),
    tabIndex: "-1"
  }, header), createElement("div", {
    className: "block-editor-editor-skeleton__body"
  }, createElement("div", {
    className: "block-editor-editor-skeleton__content",
    role: "region"
    /* translators: accessibility text for the content landmark region. */
    ,
    "aria-label": __('Editor content'),
    tabIndex: "-1"
  }, content), !!sidebar && createElement("div", {
    className: "block-editor-editor-skeleton__sidebar",
    role: "region"
    /* translators: accessibility text for the settings landmark region. */
    ,
    "aria-label": __('Editor settings'),
    tabIndex: "-1"
  }, sidebar), !!publish && createElement("div", {
    className: "block-editor-editor-skeleton__publish",
    role: "region"
    /* translators: accessibility text for the publish landmark region. */
    ,
    "aria-label": __('Editor publish'),
    tabIndex: "-1"
  }, publish)), !!footer && createElement("div", {
    className: "block-editor-editor-skeleton__footer",
    role: "region"
    /* translators: accessibility text for the footer landmark region. */
    ,
    "aria-label": __('Editor footer'),
    tabIndex: "-1"
  }, footer));
}

export default navigateRegions(EditorSkeleton);
