// @wordpress/block-editor/build-module/components/link-control/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import classnames from 'classnames';
import { noop, startsWith } from 'lodash';
/**
 * WordPress dependencies
 */

import { Button, ExternalLink, VisuallyHidden } from '@wordpress/components';
import { __, sprintf } from '@wordpress/i18n';
import { useRef, useCallback, useState, Fragment, useEffect } from '@wordpress/element';
import { safeDecodeURI, filterURLForDisplay, isURL, prependHTTP, getProtocol } from '@wordpress/url';
import { useInstanceId } from '@wordpress/compose';
import { useSelect } from '@wordpress/data';
import { focus } from '@wordpress/dom';
/**
 * Internal dependencies
 */

import LinkControlSettingsDrawer from './settings-drawer';
import LinkControlSearchItem from './search-item';
import LinkControlSearchInput from './search-input';
/**
 * Default properties associated with a link control value.
 *
 * @typedef WPLinkControlDefaultValue
 *
 * @property {string}   url           Link URL.
 * @property {string=}  title         Link title.
 * @property {boolean=} opensInNewTab Whether link should open in a new browser
 *                                    tab. This value is only assigned if not
 *                                    providing a custom `settings` prop.
 */

/**
 * Custom settings values associated with a link.
 *
 * @typedef {{[setting:string]:any}} WPLinkControlSettingsValue
 */

/**
 * Custom settings values associated with a link.
 *
 * @typedef WPLinkControlSetting
 *
 * @property {string} id    Identifier to use as property for setting value.
 * @property {string} title Human-readable label to show in user interface.
 */

/**
 * Properties associated with a link control value, composed as a union of the
 * default properties and any custom settings values.
 *
 * @typedef {WPLinkControlDefaultValue&WPLinkControlSettingsValue} WPLinkControlValue
 */

/** @typedef {(nextValue:WPLinkControlValue)=>void} WPLinkControlOnChangeProp */

/**
 * @typedef WPLinkControlProps
 *
 * @property {(WPLinkControlSetting[])=}  settings               An array of settings objects. Each object will used to
 *                                                               render a `ToggleControl` for that setting.
 * @property {boolean=}                   forceIsEditingLink     If passed as either `true` or `false`, controls the
 *                                                               internal editing state of the component to respective
 *                                                               show or not show the URL input field.
 * @property {WPLinkControlValue=}        value                  Current link value.
 * @property {WPLinkControlOnChangeProp=} onChange               Value change handler, called with the updated value if
 *                                                               the user selects a new link or updates settings.
 * @property {boolean=}                   showInitialSuggestions Whether to present initial suggestions immediately.
 */

/**
 * Renders a link control. A link control is a controlled input which maintains
 * a value associated with a link (HTML anchor element) and relevant settings
 * for how that link is expected to behave.
 *
 * @param {WPLinkControlProps} props Component props.
 */

function LinkControl(_ref) {
  var value = _ref.value,
      settings = _ref.settings,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === void 0 ? noop : _ref$onChange,
      showInitialSuggestions = _ref.showInitialSuggestions,
      forceIsEditingLink = _ref.forceIsEditingLink;
  var wrapperNode = useRef();
  var instanceId = useInstanceId(LinkControl);

  var _useState = useState(value && value.url || ''),
      _useState2 = _slicedToArray(_useState, 2),
      inputValue = _useState2[0],
      setInputValue = _useState2[1];

  var _useState3 = useState(forceIsEditingLink !== undefined ? forceIsEditingLink : !value || !value.url),
      _useState4 = _slicedToArray(_useState3, 2),
      isEditingLink = _useState4[0],
      setIsEditingLink = _useState4[1];

  var isEndingEditWithFocus = useRef(false);

  var _useSelect = useSelect(function (select) {
    var _select = select('core/block-editor'),
        getSettings = _select.getSettings;

    return {
      fetchSearchSuggestions: getSettings().__experimentalFetchLinkSuggestions
    };
  }, []),
      fetchSearchSuggestions = _useSelect.fetchSearchSuggestions;

  var displayURL = value && filterURLForDisplay(safeDecodeURI(value.url)) || '';
  useEffect(function () {
    if (forceIsEditingLink !== undefined && forceIsEditingLink !== isEditingLink) {
      setIsEditingLink(forceIsEditingLink);
    }
  }, [forceIsEditingLink]);
  useEffect(function () {
    // When `isEditingLink` is set to `false`, a focus loss could occur
    // since the link input may be removed from the DOM. To avoid this,
    // reinstate focus to a suitable target if focus has in-fact been lost.
    // Note that the check is necessary because while typically unsetting
    // edit mode would render the read-only mode's link element, it isn't
    // guaranteed. The link input may continue to be shown if the next value
    // is still unassigned after calling `onChange`.
    var hadFocusLoss = isEndingEditWithFocus.current && wrapperNode.current && !wrapperNode.current.contains(document.activeElement);

    if (hadFocusLoss) {
      // Prefer to focus a natural focusable descendent of the wrapper,
      // but settle for the wrapper if there are no other options.
      var nextFocusTarget = focus.focusable.find(wrapperNode.current)[0] || wrapperNode.current;
      nextFocusTarget.focus();
    }

    isEndingEditWithFocus.current = false;
  }, [isEditingLink]);
  /**
   * onChange LinkControlSearchInput event handler
   *
   * @param {string} val Current value returned by the search.
   */

  var onInputChange = function onInputChange() {
    var val = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    setInputValue(val);
  };

  var handleDirectEntry = function handleDirectEntry(val) {
    var type = 'URL';
    var protocol = getProtocol(val) || '';

    if (protocol.includes('mailto')) {
      type = 'mailto';
    }

    if (protocol.includes('tel')) {
      type = 'tel';
    }

    if (startsWith(val, '#')) {
      type = 'internal';
    }

    return Promise.resolve([{
      id: '-1',
      title: val,
      url: type === 'URL' ? prependHTTP(val) : val,
      type: type
    }]);
  };

  var handleEntitySearch =
  /*#__PURE__*/
  function () {
    var _ref2 = _asyncToGenerator(
    /*#__PURE__*/
    _regeneratorRuntime.mark(function _callee(val, args) {
      var results, couldBeURL;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return Promise.all([fetchSearchSuggestions(val, _objectSpread({}, args.isInitialSuggestions ? {
                perPage: 3
              } : {})), handleDirectEntry(val)]);

            case 2:
              results = _context.sent;
              couldBeURL = !val.includes(' '); // If it's potentially a URL search then concat on a URL search suggestion
              // just for good measure. That way once the actual results run out we always
              // have a URL option to fallback on.

              return _context.abrupt("return", couldBeURL && !args.isInitialSuggestions ? results[0].concat(results[1]) : results[0]);

            case 5:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function handleEntitySearch(_x, _x2) {
      return _ref2.apply(this, arguments);
    };
  }();
  /**
   * Cancels editing state and marks that focus may need to be restored after
   * the next render, if focus was within the wrapper when editing finished.
   */


  function stopEditing() {
    isEndingEditWithFocus.current = !!wrapperNode.current && wrapperNode.current.contains(document.activeElement);
    setIsEditingLink(false);
  } // Effects


  var getSearchHandler = useCallback(function (val, args) {
    var isInternal = startsWith(val, '#');
    var handleManualEntry = isInternal || isURL(val) || val && val.includes('www.');
    return handleManualEntry ? handleDirectEntry(val, args) : handleEntitySearch(val, args);
  }, [handleDirectEntry, fetchSearchSuggestions]); // Render Components

  var renderSearchResults = function renderSearchResults(_ref3) {
    var suggestionsListProps = _ref3.suggestionsListProps,
        buildSuggestionItemProps = _ref3.buildSuggestionItemProps,
        suggestions = _ref3.suggestions,
        selectedSuggestion = _ref3.selectedSuggestion,
        isLoading = _ref3.isLoading,
        isInitialSuggestions = _ref3.isInitialSuggestions;
    var resultsListClasses = classnames('block-editor-link-control__search-results', {
      'is-loading': isLoading
    });
    var manualLinkEntryTypes = ['url', 'mailto', 'tel', 'internal'];
    var searchResultsLabelId = isInitialSuggestions ? "block-editor-link-control-search-results-label-".concat(instanceId) : undefined;
    var labelText = isInitialSuggestions ? __('Recently updated') : sprintf(__('Search results for %s'), inputValue); // According to guidelines aria-label should be added if the label
    // itself is not visible.
    // See: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role

    var ariaLabel = isInitialSuggestions ? undefined : labelText;
    var SearchResultsLabel = createElement("span", {
      className: "block-editor-link-control__search-results-label",
      id: searchResultsLabelId,
      "aria-label": ariaLabel
    }, labelText);
    return createElement("div", {
      className: "block-editor-link-control__search-results-wrapper"
    }, isInitialSuggestions ? SearchResultsLabel : createElement(VisuallyHidden, null, SearchResultsLabel), createElement("div", _extends({}, suggestionsListProps, {
      className: resultsListClasses,
      "aria-labelledby": searchResultsLabelId
    }), suggestions.map(function (suggestion, index) {
      return createElement(LinkControlSearchItem, {
        key: "".concat(suggestion.id, "-").concat(suggestion.type),
        itemProps: buildSuggestionItemProps(suggestion, index),
        suggestion: suggestion,
        onClick: function onClick() {
          onChange(_objectSpread({}, value, {}, suggestion));
          stopEditing();
        },
        isSelected: index === selectedSuggestion,
        isURL: manualLinkEntryTypes.includes(suggestion.type.toLowerCase()),
        searchTerm: inputValue
      });
    })));
  };

  return createElement("div", {
    tabIndex: -1,
    ref: wrapperNode,
    className: "block-editor-link-control"
  }, isEditingLink || !value ? createElement(LinkControlSearchInput, {
    value: inputValue,
    onChange: onInputChange,
    onSelect: function onSelect(suggestion) {
      onChange(_objectSpread({}, value, {}, suggestion));
      stopEditing();
    },
    renderSuggestions: renderSearchResults,
    fetchSuggestions: getSearchHandler,
    showInitialSuggestions: showInitialSuggestions
  }) : createElement(Fragment, null, createElement("p", {
    className: "screen-reader-text",
    id: "current-link-label-".concat(instanceId)
  }, __('Currently selected'), ":"), createElement("div", {
    "aria-labelledby": "current-link-label-".concat(instanceId),
    "aria-selected": "true",
    className: classnames('block-editor-link-control__search-item', {
      'is-current': true
    })
  }, createElement("span", {
    className: "block-editor-link-control__search-item-header"
  }, createElement(ExternalLink, {
    className: "block-editor-link-control__search-item-title",
    href: value.url
  }, value && value.title || displayURL), value && value.title && createElement("span", {
    className: "block-editor-link-control__search-item-info"
  }, displayURL)), createElement(Button, {
    isSecondary: true,
    onClick: function onClick() {
      return setIsEditingLink(true);
    },
    className: "block-editor-link-control__search-item-action"
  }, __('Edit')))), createElement(LinkControlSettingsDrawer, {
    value: value,
    settings: settings,
    onChange: onChange
  }));
}

export default LinkControl;
