// @wordpress/editor/build-module/components/entities-saved-states/index.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import EquivalentKeyMap from 'equivalent-key-map';
/**
 * WordPress dependencies
 */

import { CheckboxControl, Modal, Button } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { useSelect, useDispatch } from '@wordpress/data';
import { useState } from '@wordpress/element';

var EntitiesSavedStatesCheckbox = function EntitiesSavedStatesCheckbox(_ref) {
  var id = _ref.id,
      name = _ref.name,
      rawRecord = _ref.changes.rawRecord,
      checked = _ref.checked,
      setCheckedById = _ref.setCheckedById;
  return createElement(CheckboxControl, {
    label: "".concat(name, ": \"").concat(rawRecord.name || rawRecord.slug || rawRecord.title || __('Untitled'), "\""),
    checked: checked,
    onChange: function onChange(nextChecked) {
      return setCheckedById(id, nextChecked);
    }
  });
};

export default function EntitiesSavedStates(_ref2) {
  var isOpen = _ref2.isOpen,
      _onRequestClose = _ref2.onRequestClose,
      _ref2$ignoredForSave = _ref2.ignoredForSave,
      ignoredForSave = _ref2$ignoredForSave === void 0 ? new EquivalentKeyMap() : _ref2$ignoredForSave;
  var entityRecordChangesByRecord = useSelect(function (select) {
    return select('core').getEntityRecordChangesByRecord();
  });

  var _useDispatch = useDispatch('core'),
      saveEditedEntityRecord = _useDispatch.saveEditedEntityRecord;

  var _useState = useState(function () {
    return new EquivalentKeyMap();
  }),
      _useState2 = _slicedToArray(_useState, 2),
      checkedById = _useState2[0],
      _setCheckedById = _useState2[1];

  var setCheckedById = function setCheckedById(id, checked) {
    return _setCheckedById(function (prevCheckedById) {
      var nextCheckedById = new EquivalentKeyMap(prevCheckedById);

      if (checked) {
        nextCheckedById.set(id, true);
      } else {
        nextCheckedById.delete(id);
      }

      return nextCheckedById;
    });
  };

  var saveCheckedEntities = function saveCheckedEntities() {
    checkedById.forEach(function (_checked, id) {
      if (!ignoredForSave.has(id)) {
        saveEditedEntityRecord.apply(void 0, _toConsumableArray(id.filter(function (s, i) {
          return i !== id.length - 1 || s !== 'undefined';
        })));
      }
    });

    _onRequestClose(checkedById);
  };

  return isOpen && createElement(Modal, {
    title: __('What do you want to save?'),
    onRequestClose: function onRequestClose() {
      return _onRequestClose();
    },
    contentLabel: __('Select items to save.')
  }, Object.keys(entityRecordChangesByRecord).map(function (changedKind) {
    return Object.keys(entityRecordChangesByRecord[changedKind]).map(function (changedName) {
      return Object.keys(entityRecordChangesByRecord[changedKind][changedName]).map(function (changedKey) {
        var id = [changedKind, changedName, changedKey];
        return createElement(EntitiesSavedStatesCheckbox, {
          key: id.join(' | '),
          id: id,
          name: changedName,
          changes: entityRecordChangesByRecord[changedKind][changedName][changedKey],
          checked: checkedById.get(id),
          setCheckedById: setCheckedById
        });
      });
    });
  }), createElement(Button, {
    isPrimary: true,
    disabled: checkedById.size === 0,
    onClick: saveCheckedEntities,
    className: "editor-entities-saved-states__save-button"
  }, __('Save')));
}
